### PR TITLE
fix 1q matrix bug in Quantum Shannon Decomposer

### LINKED ISSUE #10036: Quantum shannon decomposition failing for some inputs
### Environment

- **Qiskit Terra version**: 0.42.1
- **Python version**: 3.10.8

### What is happening?

qs_decomposition in [qsd.py](https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/quantum_info/synthesis/qsd.py) throws an error for some examples, seemingly due to a bug in the code

### How can we reproduce the issue?

```
from qiskit.quantum_info.synthesis.qsd import qs_decomposition
qs_decomposition(np.array([[0,1],[1,0]]))
```
Output:
```
Traceback (most recent call last):
  Cell In[14], line 3
    qs_decomposition(np.array([[0,1],[1,0]]))
  File /opt/conda/lib/python3.10/site-packages/qiskit/quantum_info/synthesis/qsd.py:122 in qs_decomposition
    return _apply_a2(circ)
  File /opt/conda/lib/python3.10/site-packages/qiskit/quantum_info/synthesis/qsd.py:253 in _apply_a2
    qc3 = two_qubit_decompose.two_qubit_cnot_decompose(mat2)
UnboundLocalError: local variable 'mat2' referenced before assignment
 
Use %tb to get the full traceback.
```
Alternatively,
```
qs_decomposition(qiskit.quantum_info.random_unitary(4).to_matrix())
```
Giving the same error

### What should happen?

We should still be able to produce a decomposed circuit from these examples

### Any suggestions?

This seems to occur when line 233 of qsd.py in the function ‘_apply_a2()’ does not transpile to include any of the ‘qsd2q’ gate type for an instance.

To fix this add something such as the following before the loop on line 242:
```
if not ind2q:
    return ccirc
```
Additionally should add a test for this kind of case to the [test](https://github.com/Qiskit/qiskit-terra/blob/main/test/python/quantum_info/test_synthesis.py)

### PR DESCRIPTION
**Pull request by ewinston**
### Summary
Fixes #10036

### CODE REVIEW

**Comment by jlapeyre** [comment]
I think this example from the bug report #10036 still fails with this commit
```python
qs_decomposition(qiskit.quantum_info.random_unitary(4).to_matrix())
```

**Comment by jlapeyre** [comment]
I added a couple of lines and a test for a matrix that still failed with the PR, and made a PR to this PR branch https://github.com/ewinston/qiskit-terra/pull/4 

The matrix I used in the test is `CXGate().to_matrix()`. Note that the qsd decomposition is more complicated than the input, but at least it does not throw an error.

**Comment by ewinston** [comment]
@jlapeyre Since the fix you and @jsmallz333 proposed works regardless of my approach maybe you should just submit that pr separately and I'll close this one.

**Comment by jlapeyre** [comment]
@ewinston , as we discussed out-of-band I think that it makes sense to include the commits you made in this PR as well. This will avoid calling the synthesis method in some cases where it can't do any good.

Unless I misunderstand, I think you can just commit https://github.com/ewinston/qiskit-terra/pull/4 in order to include the commits that fix the bug proper. If you prefer separating the PRs, I'm fine with making a separate PR as you suggest above.

**Comment by jlapeyre** [comment]
This looks good to go. But I think someone other than me should merge, since I added a commit to this PR.

File: test/python/quantum_info/test_synthesis.py [hunk 1]
```diff
@@ -1542,6 +1542,20 @@ def test_opt_a1a2(self, nqubits):
             ccirc.count_ops().get("cx"), (23 / 48) * 4**nqubits - (3 / 2) * 2**nqubits + 4 / 3
         )
 
+    # 10036
```
Comment (mtreinish) [inline test/python/quantum_info/test_synthesis.py:1545]: Nit: I think a little more context here would be useful. Just seeing  `10036` it wasn't obvious at first that this was the issue number that the test was coming from. I think just making this a link would be more useful, especially coming back to this in a year or two and we're not so fresh on the issue number being fixed here.

```suggestion
    # From: https://github.com/Qiskit/qiskit-terra/issues/10036
```

**Review by mtreinish — COMMENTED** [review]
This LGTM, I just have a small nit inline about the comments. ~I'm going to apply the suggestion and enqueue for merging.~ (I forgot about the missing release note).

**Review by mtreinish — COMMENTED** [review]
Oh, I also forgot about a release note. Would you mine adding one so we can document we fixed this in the next bugfix release.

File: test/python/quantum_info/test_synthesis.py (hunk 1 above)
Comment (jlapeyre) [reply test/python/quantum_info/test_synthesis.py:1545]: I was copying a common style for marking the bug report whose fix is being tested. I agree that with online code browsers and contemporary IDE's having it linkable would be great. In fact, GH is close to doing this automatically. It already makes the link in comments in PRs and issues. I would at least vote for a single style in qiskit (or at least terra) when people choose to mention the bug. For example always include the full url... But going further, I often ask myself "why is this test here" and see others discuss it too. And wonder if the test is made to ensure an "invariant" or even just enforce some API or design decision. It was the latter discussions that prompted me to add an issue/PR number.

**Review by mtreinish — COMMENTED** [review]
This LGTM, but I think it's still missing a release note to cover the bug fix

**Comment by jlapeyre** [comment]
[Add release note for fix to issue 10036 #5](https://github.com/ewinston/qiskit-terra/pull/5)

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for adding the release note.

**Pull request merged**